File: BOJ/1753.cpp
// 1753번_최단경로
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#define MP make_pair
#define INF 987654321
#define F_IO ios_base::sync_with_stdio(0);cin.tie(0);cout.tie(0);
using namespace std;
int v, e, k, min_distance[20001];
vector< pair<int, int> > edge[20001];
void dijkstra(int begin){
    priority_queue< pair<int, int>, vector< pair<int, int> >, greater< pair<int, int> > > pq;
    pq.push(MP(0, begin));
    min_distance[begin] = 0;
    while(!pq.empty()){
        int cur_d = pq.top().first; // 가장 가까이에 있는 정점과 그 거리
        int cur_v = pq.top().second;
        pq.pop();
        for(int i=0; i<edge[cur_v].size(); ++i){
            int next_d = min_distance[cur_v]+edge[cur_v][i].first; // 현재 정점을 거쳐 인접 정점까지 가는 거리
            int next_v = edge[cur_v][i].second; // 인접 정점
            if(min_distance[next_v] > next_d){ // 기존의 인접 정점까지의 최단거리와 현재 정점을 거쳐가는 거리를 비교
                min_distance[next_v] = next_d;
                pq.push(MP(next_d, next_v));
            }
        }
    }
}
int main(){
    F_IO;
    cin >> v >> e >> k; // 시작 정점 k
    for(int i=0; i<e; ++i){
        int u, v, w;
        cin >> u >> v >> w;
        edge[u].push_back(MP(w, v)); // u에서 v로 가는 가중치 w인 간선이 존재
    }
    fill_n(min_distance, v+1, INF);
    dijkstra(k);
    for(int i=1; i<=v; ++i){
        if(min_distance[i] == INF) cout << "INF\n";
        else cout << min_distance[i] << "\n";
    }
    return 0;
}

File: BOJ/1753.py
# 175번_최단 경로
import sys, heapq
input = sys.stdin.readline
v, e = map(int, input().split())
k = int(input()) # 시작 정점
adj = [[] for _ in range(v+1)] # 단방향 그래프 인접 리스트로 표현
for i in range(e):
    a, b, c = map(int, input().split()) # a->b 비용 c
    adj[a].append((b, c))

# 다익스트라 알고리즘으로 시작 정점에서 다른 모든 정점으로의 최단 거리를 구함
inf = 987654321
dist = [inf] * (v+1)
visited = [False] * (v+1)
def dijkstra(k):
    pq = [] # 우선순위큐로 사용할 리스트
    heapq.heappush(pq, (0, k)) # 거리, 정점번호 쌍
    dist[k] = 0 # 출발 정점의 최단 거리는 0
    while len(pq) > 0:
        cur = heapq.heappop(pq)
        cur_d = cur[0]
        cur_v = cur[1]
        for next in adj[cur_v]:
            next_v = next[0]
            next_d = next[1]
            if dist[next_v] > dist[cur_v]+next_d: # 현재 정점 거치면 더 빠른 경우
                dist[next_v] = dist[cur_v]+next_d
                visited[next_v] = True
                heapq.heappush(pq, (dist[next_v], next_v))
                
dijkstra(k) # k에서 시작
for i in range(1, v+1):
    if dist[i] == inf: print("INF")
    else: print(dist[i])